src/main/java/com/ecommerce/service/impl/OrderServiceImpl.java

```diff
@@ -68,6 +70,32 @@ public Page<OrderDTO> findOrdersByUserId(Long userId, FilterDTO filter) {
         });
     }
 
+    @Override
+    public void cancelOrder(Long id, Long userId) {
+        // set status
+        Order order = orderDAO.findById(id).orElseThrow(() -> new NotFound("Order not found"));
+        //check user
+        if (!order.getUserId().equals(userId)) {
+            throw new NotFound("User is not valid!");
+        }
+
+        if (order.getStatus().equals(Order.Status.PENDING)) {
+            order.setStatus(Order.Status.CANCELLED);
+            orderDAO.save(order);
+            // update product quantity
+            order.getOrderDetails().forEach(orderDetail -> {
+                int quantity = orderDetail.getQuantity();
+                Product product = orderDetail.getProduct();
+                int oldQuantity = product.getQuantity();
+                product.setQuantity(quantity + oldQuantity);
+                productDAO.save(product);
+            });
+        }else{
+            throw new NotFound("Order status is not valid!");
+        }
```

Comment:
```suggestion

        if(order.getStatus().notEquals(Order.Status.PENDING)) {
          throw new NotFound("Order status is not valid!");
          return;
        }
        
            order.setStatus(Order.Status.CANCELLED);
            orderDAO.save(order);
            // update product quantity
            order.getOrderDetails().forEach(orderDetail -> {
                int quantity = orderDetail.getQuantity();
                Product product = orderDetail.getProduct();
                int oldQuantity = product.getQuantity();
                product.setQuantity(quantity + oldQuantity);
                productDAO.save(product);
            });
      
```
chị suggest nên làm thế này nhé

chúng ta sẽ cố gắng trả về kết quả sớm nhất cho method

Reply:
em đã sửa rồi ạ